### PR TITLE
R22.0 redesign am

### PR DESCRIPTION
**Pull request by AijunHall**
Contributes to #51 
Removed `am` class exports in favor of interface exports. Introduced the `am/interfaces` barrel/index file to clean up exports. Made `Infill` and `Config` default interface exports, according to Typescript recommendations: 

https://www.typescriptlang.org/docs/handbook/modules.html#if-youre-only-exporting-a-single-class-or-function-use-export-default

Going forward into the `chop`, `fea`, etc.. will follow this same pattern.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,11 +1,14 @@
 {
   "name": "teton-thor",
-  "version": "22.0.3",
+  "version": "22.0.4",
   "description": "Thor API JS Library",
   "private": false,
   "main": "dist/thor.js",
   "types": "dist/thor.d.ts",
-  "exports": "./dist/thor.js",
+  "exports": {
```
Comment (AijunHall) [inline package.json:8]: This is the problem area regarding what we discussed before.

I expect the new `am` import statements in ss-gcp-ui to look like:
`import { Config, Infill } from 'teton-thor/am'`

But typescript ends up spitting this back at me 
```
Cannot find module 'teton-thor/am' or its corresponding type declarations.
```

**Review by dheaney — APPROVED** [review]

File: package.json (hunk 1 above)
Comment (bradymadams) [reply package.json:8]: @AijunHall have you learned anything new here?

Comment (AijunHall) [reply package.json:8]: @bradymadams I've setup my own test module like you mentioned to dial this in. So far I've had no luck going strictly off the npm module documentation. 

Last Friday I found these two stackoverflow posts that seem to be the way to actually fix this/do what we want. They basically just say to include a `package.json` for each submodule to expose objects without needing to specify `dist` in the import path. Trying this route today, I'll update you on how this works out.

https://stackoverflow.com/questions/58554339/typescript-npm-package-how-to-not-having-to-import-from-dist

https://stackoverflow.com/questions/55070559/npm-package-export-path-definitions

Comment (bradymadams) [reply package.json:8]: Okay, if that's the only solution we may need to weigh the advantages vs the burden of maintaining multiple package.json files.

**Comment by AijunHall** [comment]
@bradymadams I can confirm now that the submodule package json method works. `teton-thor/am` is now a valid import path.

With this solution, the new structure of thorjs would look something like:
```
thorjs
  package.json // top level json
  src
    am
      ...
      interfaces.ts
    chop
      ...
      interfaces.ts
    fea
      ...
      interfaces.ts
    smartslice
      ...
      interfaces.ts
    thor
      ...
      interfaces.ts

  // submodule jsons
  am
    package.json
  chop
    package.json
  fea
    package.json
  smartslice
    package.json
  thor
    package.json



```

**Comment by AijunHall** [comment]
> Okay, if that's the only solution we may need to weigh the advantages vs the burden of maintaining multiple package.json files.

This is the only solution I've found _so far_. I'm assuming that the nature of typescript compiling into `dist` makes the npm documentation I've been looking at not 100% applicable for us.

The sole purpose of these new multiple submodule `package.jsons` is to reference into the `dist` and present a "consumer friendly" custom path for access into what we expose through the barrel/index files

The advantages would be that things are much cleaner to import obviously since we have full control over the path strings. 

All these new directories holding the `package.json` files have to be included in the top level `package.json` files field. I don't know if that's a negative thing.

What do you think the disadvantages could be?

**Review by bradymadams — APPROVED** [review]
@AijunHall I think this looks good. My two concerns would be: 1. maintainability and 2. is it an atypical way to manage a TS library?

After looking at the sub package.json I don't think 1 is a concern anymore. Those look trivial to manage. I couldn't provide any insight on 2 - if you think this is a standard way to manage a TS library then let's go with it. Let me know your thoughts.

**Comment by AijunHall** [comment]
> I couldn't provide any insight on 2 - if you think this is a standard way to manage a TS library then let's go with it. Let me know your thoughts.

@bradymadams The use of barrel/index files is definitely a standard way to manage a javascript/typescript library. This new `package.json` method is an easy way for us to have intuitive import paths for our unique purposes. The structure of the barrel/index files for exposing interfaces, and the custom access path defined by the `package.json` work hand in hand I think.

I'm comfortable with this structure going forward, if you also think its good.

**Comment by bradymadams** [comment]
Sounds good to me. Let's go with it. I'll merge this when I get back to a computer. In the meantime, feel free to work on restructuring the rest of the modules.

**Pull request merged**